### PR TITLE
Added ButtonStyle.foregroundBuilder and ButtonStyle.backgroundBuilder

### LINKED ISSUE #139456: Allow arbitrary foreground and background elements for `ButtonStyle`
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Use case

I want to add fairly arbitrary visual details to my app's `ElevatedButton`'s (this feature request applies to all the material button types, but I'll use `ElevatedButton` as a stand in).

Here are some examples:
1. Adding a linear/radial/etc gradient
2. Adding a border
3. Using a semi-opaque image to texture the button
4. Using an arbitrary box shadow instead of the material elevation system
5. Placing .svgs above, behind and around the button as visual accents

As-is, `ButtonStyle` limits me to a very narrow range of style customization: shape, color, elevation...

I could create my own button system, but the material buttons do an enormous amount of extremely helpful things (manage splash factory, handle material state, tune visual density to match the platform, etc) that would be a huge hassle to reverse engineer.

Here are some examples of random apps in the wild that have buttons that are not achievable without heavy workarounds.

My app. I use a linear gradient and a border using a complicated work around that puts every elevated button in a stack:

Marvel snap. Marvel snap uses a particularly complex system containing tons of elements:
1. asymmetric shape
2. Texture
3. Texture fades to solid background in center
4. Stroke text
5. Metallic border
6. Light flare
7. And more...

### Proposal

I think there are a lot of different ways to add greater customization to `ButtonStyle`. I will list a few I've thought of here, I'm very open to more suggestions!
1. Mimic `Container`'s foreground and background `BoxDecoration` arguments
I don't love this solution as it's fairly limiting. For example, you can't use a `Stack` to stack elements on top of each other in this context.
2. Include foreground and background builder arguments. The builders would take in a context, the current material state(s) and the `EdgeInsets` by which the button is padded with (as they should be able to draw in and outside of the button without having to reverse engineer the complicated padding logic).
I like this much more, as it gives me completely arbitrary design control. For example, even Marvel Snap's insanely complicated button would probably be doable with this system. The implementation would need to ensure the foreground child doesn't interfere with hit testing. The disadvantage is that even pretty simple customizations (like adding a linear gradient) would be pretty complicated.

In both cases, the background argument would be placed between the button's child and the background `Material`. And/or the `Material` could be left out entirely if the background argument is non null.

Maybe to ensure that simple customization is easy while complex customization is still possible, BOTH `BoxDecoration` and builder based customization could be provided?

What this might look like to use (coded off the top of my head, probably has minor syntax errors):
```dart
ElevatedButton(
  style: ButtonStyle(
    backgroundBuilder: (context, padding, materialStates, style) {
       final baseColor = style.resolveFor(materialStates);
       final shape = style.shape;
       return Padding(
         // Pad the background by the inherited padding.
         // Elements could be placed above this padding to draw outside
         // of the elevated button's background.
         padding: padding,
         child: _ClipToShape(
           shape: shape, 
           child: Stack(
             children: [
               _MyGradient(color:  baseColor),
               _MySvgAccent(),
               // And so on.
             ],
           )
         ),
       );
    },
  ),
);
```

### PR DESCRIPTION
**Pull request by HansMuller**
Fixes https://github.com/flutter/flutter/issues/139456, https://github.com/flutter/flutter/issues/130335, https://github.com/flutter/flutter/issues/89563.

Two new properties have been added to ButtonStyle to make it possible to insert arbitrary state-dependent widgets in a button's background or foreground. These properties can be specified for an individual button, using the style parameter, or for all buttons using a button theme's style parameter.

The new ButtonStyle properties are `backgroundBuilder` and `foregroundBuilder` and their (function) types are:

```dart
typedef ButtonLayerBuilder = Widget Function(
  BuildContext context,
  Set<MaterialState> states,
  Widget? child
);
```

The new builder functions are called whenever the button is built and the `states` parameter communicates the pressed/hovered/etc state fo the button.

## `backgroundBuilder`

Creates a widget that becomes the child of the button's Material and whose child is the rest of the button, including the button's `child` parameter.  By default the returned widget is clipped to the Material's ButtonStyle.shape.

The `backgroundBuilder` can be used to add a gradient to the button's background. Here's an example that creates a yellow/orange gradient background:

opaque-gradient-bg

```dart
TextButton(
  onPressed: () {},
  style: TextButton.styleFrom(
    backgroundBuilder: (BuildContext context, Set<MaterialState> states, Widget? child) {
      return DecoratedBox(
        decoration: BoxDecoration(
          gradient: LinearGradient(colors: [Colors.orange, Colors.yellow]),
        ),
        child: child,
      );
    },
  ),
  child: Text('Text Button'),
)
```

Because the background widget becomes the child of the button's Material, if it's opaque (as it is in this case) then it obscures the overlay highlights which are painted on the button's Material. To ensure that the highlights show through one can decorate the background with an `Ink` widget.  This version also overrides the overlay color to be (shades of) red, because that makes the highlights look a little nicer with the yellow/orange background.

ink-gradient-bg

```dart
TextButton(
  onPressed: () {},
  style: TextButton.styleFrom(
    overlayColor: Colors.red,
    backgroundBuilder: (BuildContext context, Set<MaterialState> states, Widget? child) {
      return Ink(
        decoration: BoxDecoration(
          gradient: LinearGradient(colors: [Colors.orange, Colors.yellow]),
        ),
        child: child,
      );
    },
  ),
  child: Text('Text Button'),
)
```

Now the button's overlay highlights are painted on the Ink widget. An Ink widget isn't needed if the background is sufficiently translucent. This version of the example creates a translucent backround widget. 

translucent-graident-bg

```dart
TextButton(
  onPressed: () {},
  style: TextButton.styleFrom(
    overlayColor: Colors.red,
    backgroundBuilder: (BuildContext context, Set<MaterialState> states, Widget? child) {
      return DecoratedBox(
        decoration: BoxDecoration(
          gradient: LinearGradient(colors: [
            Colors.orange.withOpacity(0.5),
            Colors.yellow.withOpacity(0.5),
          ]),
        ),
        child: child,
      );
    },
  ),
  child: Text('Text Button'),
)
```

One can also decorate the background with an image. In this example, the button's background is an burlap texture image. The foreground color has been changed to black to make the button's text a little clearer relative to the mottled brown backround.

burlap-bg

```dart
TextButton(
  onPressed: () {},
  style: TextButton.styleFrom(
    foregroundColor: Colors.black,
    backgroundBuilder: (BuildContext context, Set<MaterialState> states, Widget? child) {
      return Ink(
        decoration: BoxDecoration(
          image: DecorationImage(
            image: NetworkImage(burlapUrl),
            fit: BoxFit.cover,
          ),
        ),
        child: child,
      );
    },
  ),
  child: Text('Text Button'),
)
```

The background widget can depend on the `states` parameter. In this example the blue/orange gradient flips horizontally when the button is hovered/pressed.

gradient-flip

```dart
TextButton(
  onPressed: () {},
  style: TextButton.styleFrom(
    backgroundBuilder: (BuildContext context, Set<MaterialState> states, Widget? child) {
      final Color color1 = Colors.blue.withOpacity(0.5);
      final Color color2 = Colors.orange.withOpacity(0.5);
      return DecoratedBox(
        decoration: BoxDecoration(
          gradient: LinearGradient(
            colors: switch (states.contains(MaterialState.hovered)) {
              true => <Color>[color1, color2],
              false => <Color>[color2, color1],
            },
          ),
        ),
        child: child,
      );
    },
  ),
  child: Text('Text Button'),
)
```

The preceeding examples have not included a BoxDecoration border because ButtonStyle already supports `ButtonStyle.shape` and `ButtonStyle.side` parameters that can be uesd to define state-dependent borders. Borders defined with the ButtonStyle side parameter match the button's shape. To add a border that changes color when the button is hovered or pressed, one must specify the side property using `copyWith`, since there's no `styleFrom` shorthand for this case.

border-gradient-bg

```dart
TextButton(
  onPressed: () {},
  style: TextButton.styleFrom(
    foregroundColor: Colors.indigo,
    backgroundBuilder: (BuildContext context, Set<MaterialState> states, Widget? child) {
      final Color color1 = Colors.blue.withOpacity(0.5);
      final Color color2 = Colors.orange.withOpacity(0.5);
      return DecoratedBox(
        decoration: BoxDecoration(
          gradient: LinearGradient(
            colors: switch (states.contains(MaterialState.hovered)) {
              true => <Color>[color1, color2],
              false => <Color>[color2, color1],
            },
          ),
        ),
        child: child,
      );
    },
  ).copyWith(
    side: MaterialStateProperty.resolveWith<BorderSide?>((Set<MaterialState> states) {
      if (states.contains(MaterialState.hovered)) {
        return BorderSide(width: 3, color: Colors.yellow);
      }
      return null; // defer to the default
    }),
  ),
  child: Text('Text Button'),
)
```

Although all of the examples have created a ButtonStyle locally and only applied it to one button, they could have configured the `ThemeData.textButtonTheme` instead and applied the style to all TextButtons. And, of course, all of this works for all of the ButtonStyleButton classes, not just TextButton.

## `foregroundBuilder`

Creates a Widget that contains the button's child parameter. The returned widget is clipped by the button's [ButtonStyle.shape] inset by the button's [ButtonStyle.padding] and aligned by the button's [ButtonStyle.alignment].

The `foregroundBuilder` can be used to wrap the button's child, e.g. with a border or a `ShaderMask` or as a state-dependent substitute for the child.

This example adds a border that's just applied to the child. The border only appears when the button is hovered/pressed.

border-fg

```dart
ElevatedButton(
  onPressed: () {},
  style: ElevatedButton.styleFrom(
    foregroundBuilder: (BuildContext context, Set<MaterialState> states, Widget? child) {
      final ColorScheme colorScheme = Theme.of(context).colorScheme;
      return DecoratedBox(
        decoration: BoxDecoration(
          border: states.contains(MaterialState.hovered)
            ? Border(bottom: BorderSide(color: colorScheme.primary))
            : Border(), // essentially "no border"
        ),
        child: child,
      );
    },
  ),
  child: Text('Text Button'),
)
```

The foregroundBuilder can be used with `ShaderMask` to change the way the button's child is rendered. In this example the ShaderMask's gradient causes the button's child to fade out on top.

shader_mask_fg

```dart
ElevatedButton(
  onPressed: () { },
  style: ElevatedButton.styleFrom(
    foregroundBuilder: (BuildContext context, Set<MaterialState> states, Widget? child) {
      final ColorScheme colorScheme = Theme.of(context).colorScheme;
      return ShaderMask(
        shaderCallback: (Rect bounds) {
          return LinearGradient(
            begin: Alignment.bottomCenter,
            end: Alignment.topCenter,
            colors: <Color>[
              colorScheme.primary,
              colorScheme.primaryContainer,
            ],
          ).createShader(bounds);
        },
        blendMode: BlendMode.srcATop,
        child: child,
      );
    },
  ),
  child:  const Text('Elevated Button'),
)
```

A commonly requested configuration for butttons has the developer provide images, one for pressed/hovered/normal state. You can use the foregroundBuilder to create a button that fades between a normal image and another image when the button is pressed. In this case the foregroundBuilder doesn't use the child it's passed, even though we've provided the required TextButton child parameter.

image-button

```dart
TextButton(
  onPressed: () {},
  style: TextButton.styleFrom(
    foregroundBuilder: (BuildContext context, Set<MaterialState> states, Widget? child) {
      final String url = states.contains(MaterialState.pressed) ? smiley2Url : smiley1Url;
      return AnimatedContainer(
        width: 100,
        height: 100,
        duration: Duration(milliseconds: 300),
        decoration: BoxDecoration(
          image: DecorationImage(
            image: NetworkImage(url),
            fit: BoxFit.contain,
          ),
        ),
      );
    },
  ),
  child: Text('No Child'),
)
```

In this example the button's default overlay appears when the button is hovered and pressed. Another image can be used to indicate the hovered state and the default overlay can be defeated by specifying `Colors.transparent` for the `overlayColor`:

image-per-state

```dart
TextButton(
  onPressed: () {},
  style: TextButton.styleFrom(
    overlayColor: Colors.transparent,
    foregroundBuilder: (BuildContext context, Set<MaterialState> states, Widget? child) {
      String url = states.contains(MaterialState.hovered) ? smiley3Url : smiley1Url;
      if (states.contains(MaterialState.pressed)) {
        url = smiley2Url;
      }
      return AnimatedContainer(
        width: 100,
        height: 100,
        duration: Duration(milliseconds: 300),
        decoration: BoxDecoration(
          image: DecorationImage(
            image: NetworkImage(url),
            fit: BoxFit.contain,
          ),
        ),
      );
    },
  ),
  child: Text('No Child'),
)
```

### CODE REVIEW

File: packages/flutter/lib/src/material/button_style.dart [hunk 1]
```diff
@@ -315,6 +323,22 @@ class ButtonStyle with Diagnosticable {
   /// ```
   final InteractiveInkFeatureFactory? splashFactory;
 
+  /// Creates a widget that becomes the child of the button's [Material]
+  /// and whose child is the rest of the button, including the button's
+  /// `child` parameter.
+  ///
+  /// By default the returned widget is clipped to the Material's [ButtonStyle.shape].
```
[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter/lib/src/material/button_style.dart:330]: This should probably also include a note about making sure that highlights come through by either using a non-opaque background color or the Ink widget.

Comment (goderbauer) [inline packages/flutter/lib/src/material/button_style.dart:330]: "By default" makes it sound like that this can be changed? Maybe leave a breadcrumb to tell people how this can be changed?

File: packages/flutter/lib/src/material/button_style.dart [hunk 2]
```diff
@@ -315,6 +323,22 @@ class ButtonStyle with Diagnosticable {
...
+  final ButtonLayerBuilder? backgroundBuilder;
+
+  /// Creates a Widget that contains the button's child parameter which is used
+  /// instead of the button's child.
+  ///
+  /// The returned widget is clipped by the button's
+  /// [ButtonStyle.shape] inset by the button's [ButtonStyle.padding]
+  /// and aligned by the button's [ButtonStyle.alignment].
+  /// [ButtonStyle.shape].
+  final ButtonLayerBuilder? foregroundBuilder;
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/button_style.dart:340]: After reading the two docs on foregroundBuilder and backgroundBuilder it really isn't super-clear to me what the difference would be and when to use which one (other then through trail and error). Maybe the docs can be expanded a little bit to work out the differences? There should probably also be a "See also" section for each that links to the other one?

File: packages/flutter/lib/src/material/elevated_button.dart [hunk 3]
```diff
@@ -182,9 +191,11 @@ class ElevatedButton extends ButtonStyleButton {
     final MaterialStateProperty<Color?>? foregroundColorProp = (foreground == null && disabledForeground == null)
       ? null
       : _ElevatedButtonDefaultColor(foreground, disabledForeground);
-    final MaterialStateProperty<Color?>? overlayColor = (foreground == null)
+    final MaterialStateProperty<Color?>? overlayColorProp = (foreground == null && overlayColor == null)
       ? null
-      : _ElevatedButtonDefaultOverlay(foreground);
+      : overlayColor != null && overlayColor.value == 0
+        ? const MaterialStatePropertyAll<Color?>(Colors.transparent)
+        : _ElevatedButtonDefaultOverlay((overlayColor ?? foreground)!);
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/elevated_button.dart:198]: I found this double `?:` a little hard to follow here. I wonder if a switch expression could make this more readable?

```dart
final MaterialStateProperty<Color?>? overlayColorProp = switch((foreground, overlayColor)) {
  (null, null) => null,
  (_, Color overlayColor) when overlayColor.value == 0 => const MaterialStatePropertyAll<Color?>(Colors.transparent),
  (_, _) => _ElevatedButtonDefaultOverlay((overlayColor ?? foreground)!),
};
```

(same elsewhere in this PR where the same pattern is used)

**Review by goderbauer — COMMENTED** [review]
The examples in the PR description look very cool!

Mostly just doc nits asking for more context.

File: packages/flutter/lib/src/material/elevated_button.dart (hunk 3 above)
Comment (HansMuller) [reply packages/flutter/lib/src/material/elevated_button.dart:198]: That's MUCH better. I really disliked the complicated `?:` but didn't understand how to apply the new pattern matching switch statement to them. Thanks for the pointer!

File: packages/flutter/lib/src/material/button_style.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (HansMuller) [reply packages/flutter/lib/src/material/button_style.dart:330]: Definitely planning to write more tests and to add API doc examples, yes!

Comment (HansMuller) [reply packages/flutter/lib/src/material/button_style.dart:330]: Good point, yes.

File: packages/flutter/lib/src/material/button_style.dart (hunk 2 above)
Comment (HansMuller) [reply packages/flutter/lib/src/material/button_style.dart:340]: The foreground builder inherits the same constraints as the button's child (i.e. they include the button's padding), while the background builder constraints are the same as the button's material - i.e. the button's overall bounds. I will try and make this clearer.

**Review by TahaTesser — CHANGES_REQUESTED** [review]

File: packages/flutter/lib/src/material/material_state.dart [hunk 4]
```diff
@@ -168,6 +168,9 @@ abstract class MaterialStateColor extends Color implements MaterialStateProperty
   /// specified state.
   @override
   Color resolve(Set<MaterialState> states);
+
+  /// A constant whose value is [Colors.transparent] for all states.
+  static const MaterialStateColor transparent = _MaterialStateColorTransparent();
```
Comment (TahaTesser) [inline packages/flutter/lib/src/material/material_state.dart:173]: Awesome!

Comment (HansMuller) [reply packages/flutter/lib/src/material/material_state.dart:173]: This doesn't change doesn't really belong in this PR, but I guess it can't hurt :-)

File: packages/flutter/lib/src/material/elevated_button.dart [hunk 5]
```diff
@@ -100,19 +101,26 @@ class ElevatedButton extends ButtonStyleButton {
...
+  /// If [overlayColor] is specified and its value is [Colors.transparent]
+  /// then the pressed/focused/hovered highlights are effectively defeated.
+  /// Otherwise a [MaterialStateProperty] with the same opacities as the
+  /// default is created.
   ///
   /// The [backgroundColor] and [disabledBackgroundColor] colors are
   /// used to create a [MaterialStateProperty] [ButtonStyle.backgroundColor].
   ///
+  /// Similarly, the [enabledMouseCursor] and [disabledMouseCursor]
+  /// parameters are used to construct [ButtonStyle].mouseCursor and
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/elevated_button.dart:115]: Should this be

```suggestion
  /// parameters are used to construct [ButtonStyle.mouseCursor] and
```

File: packages/flutter/lib/src/material/outlined_button.dart [hunk 6]
```diff
@@ -101,16 +101,22 @@ class OutlinedButton extends ButtonStyleButton {
...
   /// to create a [MaterialStateProperty] [ButtonStyle.foregroundColor], and
-  /// a derived [ButtonStyle.overlayColor].
+  /// a derived [ButtonStyle.overlayColor] if [overlayColor] isn't specified.
   ///
   /// The [backgroundColor] and [disabledBackgroundColor] colors are
   /// used to create a [MaterialStateProperty] [ButtonStyle.backgroundColor].
   ///
   /// Similarly, the [enabledMouseCursor] and [disabledMouseCursor]
-  /// parameters are used to construct [ButtonStyle.mouseCursor].
+  /// parameters are used to construct [ButtonStyle].mouseCursor and
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/outlined_button.dart:112]:
```suggestion
  /// parameters are used to construct [ButtonStyle.mouseCursor] and
```
?

**Review by goderbauer — APPROVED** [review]
LGTM with some doc nits.

File: examples/api/lib/material/text_button/text_button.0.dart [hunk 7]
```diff
@@ -6,78 +6,416 @@ import 'package:flutter/material.dart';
...
+    final ThemeData theme = Theme.of(context);
+    final ColorScheme colorScheme = theme.colorScheme;
+
+    // Adapt colors that are not part of the color scheme to
+    // the current dark/light mode. Used to define TextButton #7's
+    // gradients.
+    final Color color1;
+    final Color color2;
+    final Color color3;
+    if (colorScheme.brightness == Brightness.light) {
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply examples/api/lib/material/text_button/text_button.0.dart:72]: Yes!

File: examples/api/lib/material/text_button/text_button.0.dart [hunk 8]
```diff
@@ -6,78 +6,416 @@ import 'package:flutter/material.dart';
...
+                          foregroundColor: colorScheme.onError,
+                          backgroundColor: colorScheme.error,
+                        ),
+                        onPressed: () { },
+                        icon: const Icon(Icons.access_alarm),
+                        label: const Text('TextButton.icon #2'),
+                      ),
+                      verticalSpacer,
+
+                      // Override button's shape its border.
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply examples/api/lib/material/text_button/text_button.0.dart:207]: Fixed, thanks for catching that.

File: examples/api/lib/material/text_button/text_button.0.dart [hunk 9]
```diff
@@ -6,78 +6,416 @@ import 'package:flutter/material.dart';
...
+                          },
+                        ),
+                        child: const Text('TextButton #8'),
+                      ),
+                      verticalSpacer,
+
+                      // Override the foregroundBuilder to specify images for the button's pressed
+                      // hovered and inactive states.
+                      //
+                      // This is an example of completely change the default appearance of a button
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply examples/api/lib/material/text_button/text_button.0.dart:374]: Yes, fixed.

File: examples/api/lib/material/text_button/text_button.0.dart [hunk 10]
```diff
@@ -6,78 +6,416 @@ import 'package:flutter/material.dart';
...
+                      // by specifying images for each state and by turning off the overlays by
+                      // overlayColor: Colors.transparent. AnimatedContainer takes care of the
+                      // fade in and out segues between images.
+                      //
+                      // The foregroundBuilder its child parameter. Unfortunately TextButton's child
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply examples/api/lib/material/text_button/text_button.0.dart:379]: Should have been ` // This foregroundBuilder function ignores its child parameter.`. Fixed. Thanks for catching that.

File: packages/flutter/lib/src/material/button_style.dart [hunk 11]
```diff
@@ -315,6 +323,42 @@ class ButtonStyle with Diagnosticable {
...
+  ///    child and is layered behind the child.
+  ///  * [ButtonStyleButton.clipBehavior], for more information about
+  ///    configuring clipping.
+  final ButtonLayerBuilder? backgroundBuilder;
+
+  /// Creates a Widget that contains the button's child parameter which is used
+  /// instead of the button's child.
+  ///
+  /// The returned widget is clipped by the button's
+  /// [ButtonStyle.shape] inset by the button's [ButtonStyle.padding]
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply packages/flutter/lib/src/material/button_style.dart:351]: I'd meant `and`. Thanks and fixed.

File: packages/flutter/lib/src/material/elevated_button.dart (hunk 5 above)
Comment (HansMuller) [reply packages/flutter/lib/src/material/elevated_button.dart:115]: Yes, fixed.

File: packages/flutter/test/material/text_button_test.dart [hunk 12]
```diff
@@ -1925,6 +1925,203 @@ void main() {
...
+              onPressed: () {},
+              focusNode: focusNode,
+              child: const Text('button'),
+            ),
+          ),
+        ),
+      ),
+    );
+
+    // Default
```
Comment (TahaTesser) [inline packages/flutter/test/material/text_button_test.dart:2086]:
```suggestion
    // Default.
```

**Review by TahaTesser — CHANGES_REQUESTED** [review]

File: examples/api/lib/material/text_button/text_button.0.dart [hunk 13]
```diff
@@ -6,78 +6,417 @@ import 'package:flutter/material.dart';
...
+    switch (colorScheme.brightness) {
+      case Brightness.light:
+        color1 = Colors.blue.withOpacity(0.5);
+        color2 = Colors.orange.withOpacity(0.5);
+        color3 = Colors.yellow.withOpacity(0.5);
+      case Brightness.dark:
+        color1 = Colors.purple.withOpacity(0.5);
+        color2 = Colors.cyan.withOpacity(0.5);
+        color3 = Colors.yellow.withOpacity(0.5);
+    }
```
Comment (goderbauer) [inline examples/api/lib/material/text_button/text_button.0.dart:81]:
```suggestion
    final (Color color1, Color color2, Color color3) = switch (colorScheme.brightness) {
      Brightness.light => (
        Colors.blue.withOpacity(0.5),
        Colors.orange.withOpacity(0.5),
        Colors.yellow.withOpacity(0.5),
      ),
      Brightness.dark => (
        Colors.purple.withOpacity(0.5),
        Colors.cyan.withOpacity(0.5),
        Colors.yellow.withOpacity(0.5),
      ),
    };
```

File: examples/api/lib/material/text_button/text_button.0.dart [hunk 14]
```diff
@@ -6,78 +6,417 @@ import 'package:flutter/material.dart';
...
+        gradient: LinearGradient(
+          colors: switch (states.contains(MaterialState.hovered)) {
+            true => <Color>[color1, color2],
+            false => <Color>[color2, color1],
+          },
+        ),
+      );
+    }
+
+    return Row(
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply examples/api/lib/material/text_button/text_button.0.dart:102]: I added a horizontal scrollbar.

File: packages/flutter/test/material/elevated_button_test.dart [hunk 15]
```diff
@@ -1942,6 +1942,203 @@ void main() {
...
+              onPressed: () {},
+              focusNode: focusNode,
+              child: const Text('button'),
+            ),
+          ),
+        ),
+      ),
+    );
+
+    // Default
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply packages/flutter/test/material/elevated_button_test.dart:2103]: Fixed this in all of the *_button_test.dart files.

File: examples/api/lib/material/text_button/text_button.0.dart (hunk 14 above)
[the remark this answers is not in the document]
Comment (HansMuller) [reply examples/api/lib/material/text_button/text_button.0.dart:102]: When you ran the example, the image button (the last one, with the "This child is not..." label) didn't show up.  Do the ugly URLs at the bottom of the example work for you?

File: examples/api/lib/material/text_button/text_button.0.dart (hunk 13 above)
Comment (HansMuller) [reply examples/api/lib/material/text_button/text_button.0.dart:81]: This is _so_ much better. It's hard (It was hard for me) to find an example of this syntax in action. Hopefully we're helping spread the word.

File: examples/api/lib/material/text_button/text_button.0.dart (hunk 14 above)
[the remark this answers is not in the document]
Comment (TahaTesser) [reply examples/api/lib/material/text_button/text_button.0.dart:107]: I see this:

Screenshot 2024-01-30 at 21 23 23


About the urls, we could add images for this into assets repo.

We have some examples that do this, e.g https://github.com/flutter/flutter/blob/1daac1b8752fbef5eb11b47b7828837a062faf0c/examples/api/lib/material/flexible_space_bar/flexible_space_bar.0.dart#L39

File: examples/api/lib/material/text_button/text_button.0.dart [hunk 16]
```diff
@@ -6,78 +6,441 @@ import 'package:flutter/material.dart';
...
+                  const SizedBox(height: 16),
+                  Row(
+                    children: <Widget>[
+                      const Expanded(child: Text('RTL Text')),
+                      const SizedBox(width: 4),
+                      Switch(
+                        value: textDirection == TextDirection.rtl,
+                        onChanged: (bool value) {
+                          setState(() {
+                            textDirection = value ? TextDirection.rtl : TextDirection.ltr;
```
Comment (TahaTesser) [inline examples/api/lib/material/text_button/text_button.0.dart:141]: This controls card takes more vertical space than required. 

We could use  a `BottomAppBar` for controls, see this AppBar [example](https://main-api.flutter.dev/flutter/material/AppBar-class.html#material.AppBar.2).

Comment (HansMuller) [reply examples/api/lib/material/text_button/text_button.0.dart]: I like the layout ... maybe you mean `horizontal space`? If so, check out the new version. It's packed a little more tightly.

Comment (TahaTesser) [reply examples/api/lib/material/text_button/text_button.0.dart]: Great. Much more compact now.

File: examples/api/lib/material/text_button/text_button.0.dart [hunk 17]
```diff
@@ -6,78 +6,473 @@ import 'package:flutter/material.dart';
 
 /// Flutter code sample for [TextButton].
 
-void main() => runApp(const TextButtonExampleApp());
+void main() {
+  runApp(const TextButtonExampleApp());
```
Comment (TahaTesser) [inline examples/api/lib/material/text_button/text_button.0.dart:10]: An example test is needed
https://github.com/flutter/flutter/tree/master/examples/api#writing-tests

**Review by TahaTesser — CHANGES_REQUESTED** [review]

File: examples/api/lib/material/text_button/text_button.0.dart (hunk 14 above)
[the remark this answers is not in the document]
Comment (HansMuller) [reply examples/api/lib/material/text_button/text_button.0.dart:357]: That's an excellent idea: https://github.com/flutter/assets-for-api-docs/pull/235

File: examples/api/lib/material/text_button/text_button.0.dart (hunk 17 above)
Comment (HansMuller) [reply examples/api/lib/material/text_button/text_button.0.dart:10]: I was just adding one when I discovered that the testing framework doesn't like my bonkers URLs. Adding the images to the asssets-for-api-docs repo now, then I'll upload a simple test.

Comment (HansMuller) [reply examples/api/lib/material/text_button/text_button.0.dart:10]: Done.

File: examples/api/lib/material/text_button/text_button.0.dart [hunk 18]
```diff
@@ -6,78 +6,462 @@ import 'package:flutter/material.dart';
...
+        style: TextButton.styleFrom(
+          overlayColor: Colors.yellow,
+        ),
+        onPressed: () { },
+        child: const Text('TextButton #4'),
+      ),
+    ];
+
+
+    final List<Widget> columnTwoButtons = <Widget>[
```
Comment (TahaTesser) [inline examples/api/lib/material/text_button/text_button.0.dart:190]:
```suggestion
    ];

    final List<Widget> columnTwoButtons = <Widget>[
```

**Review by TahaTesser — APPROVED** [review]
LGTM!

File: examples/api/lib/material/text_button/text_button.0.dart (hunk 18 above)
Comment (HansMuller) [reply examples/api/lib/material/text_button/text_button.0.dart:190]: 👍

File: examples/api/lib/material/text_button/text_button.0.dart [hunk 19]
```diff
@@ -6,78 +6,462 @@ import 'package:flutter/material.dart';
...
+  @override
+  void initState() {
+    scrollController = ScrollController();
+    super.initState();
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+    scrollController.dispose();
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply examples/api/lib/material/text_button/text_button.0.dart:71]: 👍

**Comment by quyenlv-unicloud** [comment]
Hi, I just want to add helpful information in case someone have problem when use `backgroundBuilder` with FilledButton.

Because FilledButton default clipBehavior Clip.none so `backgroundBuilder` with color will overlap your button background, not clip.

You can use this code to apply Material shape
```
backgroundBuilder: (context, states, child) {
  return Builder(
    builder: (context) {
      final shape =
          LookupBoundary.findAncestorWidgetOfExactType<Material>(
            context,
          )?.shape;
      return Ink(
        decoration: ShapeDecoration(
          shape: shape ?? RoundedRectangleBorder(),
          gradient: LinearGradient(colors: [Colors.orange, Colors.yellow]),
        ),
        child: child,
      );
    },
  );
},
```
